File: src/android/Linphone.java
package cordova.scrachx.linphone;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.view.LayoutInflater;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.linphone.core.LinphoneAddress;
import org.linphone.core.LinphoneAuthInfo;
import org.linphone.core.LinphoneCall;
import org.linphone.core.LinphoneCallParams;
import org.linphone.core.LinphoneCoreFactory;
import org.linphone.core.LinphoneCoreException;
import org.linphone.core.LinphoneProxyConfig;
import org.linphone.core.LinphoneChatRoom;
import org.linphone.core.LinphoneChatMessage;
import org.linphone.mediastream.video.capture.hwconf.AndroidCameraConfiguration;

public class Linphone extends CordovaPlugin {

    private static LinphoneProxyConfig mProxyCfg;
    private static String mDefaultDomain;
    private static String mDefaultStun;
    private static Context mContext;
    private static LinphoneManager mLinphoneManager;
    private static LinphoneChatRoom mChatRoom;
    private static SharedPreferences mSettings;
    private static SharedPreferences.Editor mSettingsEditor;
    private static final String PREFS_NAME = "MyPrefsFile";
    private static Resources mResources;
    private static String mPackageName;
    private static LayoutInflater mLayoutInflator;

    @Override
    protected void pluginInitialize() {
        super.pluginInitialize();
        mDefaultDomain = "sip.linphone.org";
        mDefaultStun = "stun.linphone.org";
    }

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        mContext = cordova.getActivity().getApplicationContext();
        mSettings = mContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        mSettingsEditor = mSettings.edit();
        mLinphoneManager = new LinphoneManager(mContext);
        mLinphoneManager.createAndStart();
        if (mResources == null)
            mResources = cordova.getActivity().getApplication().getResources();
        if (mPackageName == null)
            mPackageName = cordova.getActivity().getApplication().getPackageName();
        if (mLayoutInflator == null) {
            mLayoutInflator = cordova.getActivity().getLayoutInflater();
        }
    }

    public boolean execute(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        if(action.equals("saveAuthInfo")) {
            saveAuthInfo(args.getString(0), args.getString(1), args.getString(2), args.getString(3), callbackContext);
            return true;
        } else if(action.equals("saveProxyInfo")) {
            saveProxyInfo(args.getString(0), args.getString(1), args.getInt(2), args.getBoolean(3), args.getBoolean(4), args.getBoolean(5), callbackContext);
            return true;
        } else if(action.equals("saveSipInfo")) {
            saveSipInfo(args.getInt(0), args.getBoolean(1), args.getInt(2), args.getInt(3), callbackContext);
            return true;
        } else if (action.equals("login")) {
            cordova.getThreadPool().execute(new Runnable() {public void run() {
                try {
                    login(callbackContext);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
            });
            return true;
        } else if (action.equals("logout")) {
            logout(callbackContext);
            return true;
        } else if (action.equals("call")) {
            cordova.getThreadPool().execute(new Runnable() {public void run() {
                    try {
                        call(args.getString(0), args.getString(1), callbackContext);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
            });
            return true;
        } else if (action.equals("sendMessage")) {
            cordova.getThreadPool().execute(new Runnable() {public void run() {
                    try {
                        sendMessage(args.getString(0), args.getString(1), args.getBoolean(2), callbackContext);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
            });
            return true;
        } else if (action.equals("answerCall")) {
            cordova.getThreadPool().execute(new Runnable() {public void run() {
                try {
                    answerCall(callbackContext);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
            });
            return true;
        } else if (action.equals("videoCall")) {
            cordova.getThreadPool().execute(new Runnable() {public void run() {
                try {
                    videoCall(args.getString(0), args.getString(1), callbackContext);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
            });
            return true;
        } else if(action.equals("hangup")){
            cordova.getThreadPool().execute(new Runnable() {public void run() {
                try {
                    hangup(callbackContext);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
            });
            return true;
        } else if(action.equals("toggleVideo")){
            toggleVideo(callbackContext);
            return true;
        } else if(action.equals("toggleSpeaker")){
            toggleSpeaker(callbackContext);
            return true;
        } else if(action.equals("toggleMute")){
            toggleMute(callbackContext);
            return true;
        } else if(action.equals("sendDtmf")){
            sendDtmf(args.getString(0), callbackContext);
            return true;
        } else if(action.equals("registerCallback")) {
            mLinphoneManager.setCallbackRegister(callbackContext);
            return true;
        } else if(action.equals("authInfoReqCallback")) {
            mLinphoneManager.setCallbackAuthInfoReq(callbackContext);
            return true;
        } else if(action.equals("globalCallback")) {
            mLinphoneManager.setCallbackGlobal(callbackContext);
            return true;
        } else if(action.equals("callStateCallback")) {
            mLinphoneManager.setCallbackCallState(callbackContext);
            return true;
        } else if(action.equals("callStatsCallback")) {
            mLinphoneManager.setCallbackCallStats(callbackContext);
            return true;
        } else if(action.equals("messageReceivedCallback")) {
            mLinphoneManager.setCallbackMessageReceived(callbackContext);
            return true;
        }
        return false;
    }

    /**https://developer.android.com/guide/topics/data/data-storage.html
     * Save and update sip info in Shared Preferences
     * @param pPort int
     * @param pIpv6 Boolean
     * @param pIncTimeOut int
     * @param pKeepAlive int
     * @param callbackContext CallbackContext
     */
    public static synchronized void saveSipInfo(final int pPort, final Boolean pIpv6, final int pIncTimeOut, final int pKeepAlive, final CallbackContext callbackContext) {
        try {
            if(pPort != 0 && pIpv6 != null && pIncTimeOut != 0 && pKeepAlive != 0) {
                mSettingsEditor.putInt("sipPort", pPort);
                mSettingsEditor.putBoolean("sipIPV6", pIpv6);
                mSettingsEditor.putInt("sipIncTimeOut", pIncTimeOut);
                mSettingsEditor.putInt("sipKeepAlive", pKeepAlive);
                mSettingsEditor.apply();
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Sip parameters saved/updated"));
            } else {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error",  "Sip parameters are empty"));
            }
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
        }
    }

    /**
     * Save and update auth info in Shared Preferences
     * @param pUsername String
     * @param pUserId String
     * @param pRealm String
     * @param pPaswd String
     * @param callbackContext CallbackContext
     */
    public static synchronized void saveAuthInfo(final String pUsername, final String pUserId, final String pRealm, final String pPaswd, final CallbackContext callbackContext) {
        try {
            if(!pUsername.isEmpty() && !pUserId.isEmpty() && !pPaswd.isEmpty()) {
                mSettingsEditor.putString("authUsername", pUsername);
                mSettingsEditor.putString("authUserId", pUserId);
                mSettingsEditor.putString("authPaswd", pPaswd);
                if(pRealm.isEmpty()) {
                    mSettingsEditor.putString("authRealm", mDefaultDomain);
                } else {
                    mSettingsEditor.putString("authRealm", pRealm);
                }
                mSettingsEditor.apply();
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Authentication parameters saved/updated"));
            } else {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Authentication parameters are empty"));
            }
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
        }
    }

    /**
     * Save and update proxy info in Shared Preferences
     * @param pRegProxy String
     * @param pRegIdentity String
     * @param pRegExpires int
     * @param pRegSendRegister Boolean
     * @param pPublish Boolean
     * @param pDialEscapePlus Boolean
     * @param callbackContext CallbackContext
     */
    public static synchronized void saveProxyInfo(final String pRegProxy, final String pRegIdentity, final int pRegExpires, final Boolean pRegSendRegister, final Boolean pPublish, final Boolean pDialEscapePlus, final CallbackContext callbackContext) {
        try {
            if(!pRegProxy.isEmpty() && !pRegIdentity.isEmpty() && pRegExpires != 0 && pRegSendRegister != null && pPublish != null && pDialEscapePlus != null) {
                mSettingsEditor.putString("regProxy", pRegProxy);
                mSettingsEditor.putString("regIdentity", pRegIdentity);
                mSettingsEditor.putInt("regExpires", pRegExpires);
                mSettingsEditor.putBoolean("regSendRegister", pRegSendRegister);
                mSettingsEditor.putBoolean("publish", pPublish);
                mSettingsEditor.putBoolean("dialEscapePlus", pDialEscapePlus);
                mSettingsEditor.apply();
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Proxy parameters saved/updated"));
            } else {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Proxy parameters are empty"));
            }
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
        }
    }

    /**
     * Log into the sip server, trigger authInfoRequested and registrationState function
     * @param callbackContext CallbackContext
     */
    public static synchronized void login(final CallbackContext callbackContext) throws JSONException {
        try {
            String username = LinphoneUtils.getDisplayableUsernameFromAddress(mSettings.getString("authUsername", "default"));
            String domain = LinphoneUtils.getDisplayableUsernameFromAddress(mSettings.getString("authRealm", "default"));
            String identity = "sip:" + username + "@" + domain;

            mProxyCfg = mLinphoneManager.getLc().createProxyConfig();
            LinphoneAuthInfo info;
            info = LinphoneCoreFactory.instance().createAuthInfo(username, mSettings.getString("authPaswd", "default"), null, domain);
            //mLinphoneManager.setStunServer(mDefaultStun);
            mLinphoneManager.getLc().addAuthInfo(info);
            mProxyCfg.setIdentity(identity);
            mProxyCfg.setProxy(domain);
            mProxyCfg.enableRegister(true);
            mLinphoneManager.getLc().addProxyConfig(mProxyCfg);
            mLinphoneManager.getLc().setDefaultProxyConfig(mProxyCfg);
            callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Logging..."));
        } catch (Exception e) {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
        }
    }

    /**
     * Log out the sip server, trigger authInfoRequested and registrationState function
     * @param callbackContext CallbackContext
     */
    public static synchronized void logout(final CallbackContext callbackContext) throws JSONException {
        try {
            mLinphoneManager.getLc().setDefaultProxyConfig(mProxyCfg);
            mProxyCfg.edit();
            mProxyCfg.enableRegister(false);
            mProxyCfg.done();
            callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Disconnecting..."));
        } catch (Exception e) {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
        }
    }

    /**
     * Log out the sip server, trigger authInfoRequested and registrationState function
     * @param pMessage String
     * @param pSipUri String
     * @param pNewChatConversation boolean
     * @param callbackContext CallbackContext
     */
    public static synchronized void sendMessage(final String pMessage, final String pSipUri, final boolean pNewChatConversation, final CallbackContext callbackContext) throws JSONException {
        try {
		boolean isNetworkReachable = mLinphoneManager.getLc() == null ? false : mLinphoneManager.getLc().isNetworkReachable();
		LinphoneAddress lAddress = null;

		if(pNewChatConversation && mChatRoom == null) {
			if (pSipUri != null && !pSipUri.equals("")) {
				lAddress = mLinphoneManager.getLc().interpretUrl(pSipUri);
				mChatRoom = mLinphoneManager.getLc().getChatRoom(lAddress);
			}
		}
		if (mChatRoom != null && pMessage != null && pMessage.length() > 0 && isNetworkReachable) {
			LinphoneChatMessage message = mChatRoom.createLinphoneChatMessage(pMessage);
			mChatRoom.sendChatMessage(message);
			// TODO : listener message state
			callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Sending..."));
		} 
        } catch (Exception e) {
            	callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
        }
    }

    /**
     * Outgoing call an sip address without video
     * @param pAddress String
     * @param pDisplayName String
     * @param callbackContext CallbackContext
     */
    public static synchronized void call(final String pAddress, final String pDisplayName, final CallbackContext callbackContext) throws JSONException {
        LinphoneAddress lAddress = null;
        try {
            lAddress = mLinphoneManager.getLc().interpretUrl(pAddress);
            LinphoneProxyConfig lpc = mLinphoneManager.getLc().getDefaultProxyConfig();
            if (lpc != null && lAddress.asStringUriOnly().equals(lpc.getIdentity())) {
                return;
            }
        } catch (LinphoneCoreException e) {
            callbackContext.error(e.getMessage());
        }
        if(!pDisplayName.isEmpty()) {
          lAddress.setDisplayName(pDisplayName);
        } else {
          lAddress.setDisplayName(LinphoneUtils.getDisplayableUsernameFromAddress(pAddress));
        }
        boolean isLowBandwidthConnection = !LinphoneUtils.isHighBandwidthConnection(mContext);
        if (mLinphoneManager.getLc().isNetworkReachable()) {
            try {
                LinphoneCallParams params = mLinphoneManager.getLc().createCallParams(null);
                params.enableLowBandwidth(isLowBandwidthConnection);
                mLinphoneManager.getLc().inviteAddressWithParams(lAddress, params);
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Calling..."));
            } catch (LinphoneCoreException e) {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
            }
        } else {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Network unreachable"));
        }
    }

    /**
     * Answer incoming call
     * @param callbackContext CallbackContext
     */
    public static synchronized void answerCall(final CallbackContext callbackContext) throws JSONException {
        if (mLinphoneManager.getLc().getCurrentCall() != null && mLinphoneManager.getLc().getCurrentCall().getState() == LinphoneCall.State.IncomingReceived) {
            try {
                mLinphoneManager.getLc().acceptCall(mLinphoneManager.getLc().getCurrentCall());
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Take call"));
            } catch (LinphoneCoreException e) {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
            }
        } else {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Not calling"));
        }
    }

    /**
     * Terminate call
     * @param callbackContext CallbackContext
     */
    public static synchronized void hangup(final CallbackContext callbackContext) throws JSONException {
        if(mLinphoneManager.getLc().getCurrentCall() != null) {
            try {
                if(mLinphoneManager.isRinging()) mLinphoneManager.stopRinging();
                mLinphoneManager.getLc().terminateCall(mLinphoneManager.getLc().getCurrentCall());
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Terminate call"));
            } catch (Exception e) {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
            }
        } else {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Not calling"));
        }
    }

    /**
     * Toggle speaker
     * @param callbackContext CallbackContext
     */
    public static synchronized void toggleSpeaker(final CallbackContext callbackContext) throws JSONException {
        if(mLinphoneManager.getLc().getCurrentCall() != null && mLinphoneManager.getLc().getCurrentCall().getState() == LinphoneCall.State.StreamsRunning) {
            try {
                mLinphoneManager.toggleEnableSpeaker();
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Speaker ON/OFF"));
            } catch (Exception e) {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
            }
        } else {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Not calling"));
        }
    }

    /**
     * Toggle mute mic
     * @param callbackContext CallbackContext
     */
    public static synchronized void toggleMute(final CallbackContext callbackContext) throws JSONException {
        if(mLinphoneManager.getLc().getCurrentCall() != null && mLinphoneManager.getLc().getCurrentCall().getState() == LinphoneCall.State.StreamsRunning) {
            try {
                mLinphoneManager.toggleMute();
                callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Mic ON/OFF"));
            } catch (Exception e) {
                callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
            }
        } else {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", "Not calling"));
        }
    }

    /**
     * Send DTMF
     * @param number String cast to char
     * @param callbackContext CallbackContext
     */
    public static synchronized void sendDtmf(final String character, final CallbackContext callbackContext) throws JSONException {
        try {
            mLinphoneManager.sendDtmf(character.charAt(0));
            callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Dtmf sent : " + number));
        } catch (Exception e) {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
        }
    }

    /**
     * Toggle video
     * @param number String cast to char
     * @param callbackContext CallbackContext
     */
    public static synchronized void toggleVideo(final CallbackContext callbackContext) throws JSONException {
        try {
            // TODO
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
        }
    }

    public static synchronized void videoCall(final String address, final String displayName, final CallbackContext callbackContext) throws JSONException {
        LinphoneAddress lAddress = null;
        try {
            int camId = 0;
            AndroidCameraConfiguration.AndroidCamera[] cameras = AndroidCameraConfiguration.retrieveCameras();
            for (AndroidCameraConfiguration.AndroidCamera androidCamera : cameras) {
                if (androidCamera.frontFacing)
                    camId = androidCamera.id;
            }
            lAddress = mLinphoneManager.getLc().interpretUrl(address);
            boolean isLowBandwidthConnection = !LinphoneUtils.isHighBandwidthConnection(mContext);
            LinphoneCallParams params = mLinphoneManager.getLc().createCallParams(null);
            params.setVideoEnabled(true);
            if(isLowBandwidthConnection) {
                params.enableLowBandwidth(true);
            }
            mLinphoneManager.getLc().setVideoDevice(camId);
            mLinphoneManager.getLc().inviteAddressWithParams(lAddress, params);
            callbackContext.success(LinphoneUtils.formatCallbackAnswer("success", "Calling with video..."));

            // TODO
            //Intent intent = new Intent(mContext, LinphoneCallVideoActivity.class);
            //mContext.startActivity(intent);
        } catch (Exception e) {
            callbackContext.error(LinphoneUtils.formatCallbackAnswer("error", e.getMessage()));
        }
    }

}

File: src/ios/Linphone.swift
import Foundation

@objc(Linphone) class Linphone : CDVPlugin {
    
    var mDefaultDomain : String = "sip.linphone.org"
    
    override init() {
    }
    
    override func pluginInitialize() {
        
    }

    @objc(saveAuthInfo:) func saveAuthInfo(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let pUsername = command.arguments[0] as? String ?? ""
        let pUserId = command.arguments[1] as? String ?? ""
        let pRealme = command.arguments[2] as? String ?? ""
        let pPassword = command.arguments[3] as? String ?? ""
        
        self.commandDelegate.run(inBackground: {
            if !pUsername.isEmpty && !pUserId.isEmpty && !pPassword.isEmpty {
                UserDefaults.standard.set(pUsername, forKey: "authUsername")
                UserDefaults.standard.set(pUserId, forKey: "authUserId")
                UserDefaults.standard.set(pPassword, forKey: "authPassword")
                if !pRealme.isEmpty {
                    UserDefaults.standard.set(pRealme, forKey: "authRealm")
                } else {
                    UserDefaults.standard.set(self.mDefaultDomain, forKey: "authRealm")
                }
                UserDefaults.standard.synchronize()
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Authentication parameters saved/updated"
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Authentication parameters are empty"
                )
            }
        
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(saveSipInfo:) func saveSipInfo(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let pPort = command.arguments[0] as? Int ?? 0
        let pIpv6 = command.arguments[1] as? Bool ?? false
        let pIncTimeout = command.arguments[2] as? Int ?? 0
        let pKeepAlive = command.arguments[3] as? Int ?? 0
        
        self.commandDelegate.run(inBackground: {
            if pPort != 0 && pIpv6 != false && pIncTimeout != 0 && pKeepAlive != 0 {
                UserDefaults.standard.set(pPort, forKey: "sipPort")
                UserDefaults.standard.set(pIpv6, forKey: "supIpv6")
                UserDefaults.standard.set(pIncTimeout, forKey: "sipIncTimeOut")
                UserDefaults.standard.set(pKeepAlive, forKey: "sipKeepAlive")
                UserDefaults.standard.synchronize()
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Sip parameters saved/updated"
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Sip parameters are empty"
                )
            }
        
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(saveProxyInfo:) func saveProxyInfo(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let pRegProxy = command.arguments[0] as? String ?? ""
        let pRegIdentity = command.arguments[1] as? String ?? ""
        let pRegExpires = command.arguments[2] as? String ?? ""
        let pRegSendRegister = command.arguments[3] as? String ?? ""
        let pPublish = command.arguments[4] as? Bool ?? true
        let pDialEscapePlus = command.arguments[5] as? Bool ?? false
        
        self.commandDelegate.run(inBackground: {
            if !pRegProxy.isEmpty && !pRegIdentity.isEmpty && pRegExpires.isEmpty && pRegSendRegister.isEmpty  {
                UserDefaults.standard.set(pRegProxy, forKey: "regProxy")
                UserDefaults.standard.set(pRegIdentity, forKey: "regIdentity")
                UserDefaults.standard.set(pRegExpires, forKey: "regExpires")
                UserDefaults.standard.set(pRegSendRegister, forKey: "regSendRegister")
                UserDefaults.standard.set(pPublish, forKey: "publish")
                UserDefaults.standard.set(pDialEscapePlus, forKey: "dialEscapePlus")
                UserDefaults.standard.synchronize()
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Proxy parameters saved/updated"
                )
            } else {
                 pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Proxy parameters are empty"
                )
            }
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(login:) func login(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        self.commandDelegate.run(inBackground: {
            let lc : OpaquePointer = LinphoneManager.sharedInstance.getLinphoneCore()
            if LinphoneUtils.isNetworkReachable() {
                let usernameSaved: String = UserDefaults.standard.object(forKey: "authUsername") as? String ?? String()
                let domainSaved: String = UserDefaults.standard.object(forKey: "authRealm") as? String ?? String()
                let passwordSaved: String = UserDefaults.standard.object(forKey: "authPassword") as? String ?? String()
                let username = LinphoneUtils.getDisplayableUsernameFromAddress(sipAddress: usernameSaved)
                //let domain = LinphoneUtils.getAddressDisplayName(uri: domainSaved)
                let identity : String = "sip:" + username + "@" + domainSaved
                let from = linphone_address_new(identity)
                
                let proxy_cfg = linphone_proxy_config_new()
                linphone_proxy_config_edit(proxy_cfg)
                let info = linphone_auth_info_new(linphone_address_get_username(from), nil, passwordSaved, nil, nil, linphone_address_get_domain(from))
                linphone_core_clear_all_auth_info(lc)
                linphone_core_add_auth_info(lc, info)
                
                linphone_proxy_config_set_identity_address(proxy_cfg, from)
                linphone_proxy_config_set_server_addr(proxy_cfg, linphone_address_get_domain(from))
                linphone_proxy_config_enable_register(proxy_cfg, 1)
                linphone_proxy_config_enable_publish(proxy_cfg, 1)
                linphone_core_add_proxy_config(lc, proxy_cfg)
                linphone_core_set_default_proxy_config(lc, proxy_cfg)
                
                linphone_proxy_config_done(proxy_cfg)
                linphone_address_destroy(from)
                
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Logging..."
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Network unreachable"
                )
            }
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(logout:) func logout(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        self.commandDelegate.run(inBackground: {
            let lc : OpaquePointer = LinphoneManager.sharedInstance.getLinphoneCore()
            let proxy_cfg = linphone_core_get_default_proxy_config(lc);
            if proxy_cfg != nil {
                linphone_proxy_config_edit(proxy_cfg);
                linphone_proxy_config_enable_register(proxy_cfg, 0);
                linphone_proxy_config_done(proxy_cfg);
                
                if linphone_proxy_config_get_state(proxy_cfg) !=  LinphoneRegistrationCleared {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_OK,
                        messageAs: "Lougout..."
                    )
                }
            }

            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(call:) func call(command: CDVInvokedUrlCommand) {
        
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let pAddress = command.arguments[0] as? String ?? ""
        //let pDisplayName = command.arguments[1] as? String ?? ""

        self.commandDelegate.run(inBackground: {
            let lc : OpaquePointer = LinphoneManager.sharedInstance.getLinphoneCore()
            var call: OpaquePointer;
            
            if !pAddress.isEmpty {
                call = linphone_core_invite(lc, pAddress)
                linphone_call_ref(call)
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Calling..."
                )
                
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Error calling..."
                )
            }
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(answerCall:) func answerCall(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        self.commandDelegate.run(inBackground: {
            let lc : OpaquePointer = LinphoneManager.sharedInstance.getLinphoneCore()
            let currentCall: OpaquePointer = linphone_core_get_current_call(lc)
            if linphone_core_get_calls_nb(lc) > 0 && linphone_call_get_state(currentCall) == LinphoneCallIncomingReceived {
                linphone_core_accept_call(lc, currentCall)
                linphone_call_ref(currentCall)
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Take call..."
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Not calling"
                )
            }
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(hangup:) func hangup(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        self.commandDelegate.run(inBackground: {
            let lc : OpaquePointer = LinphoneManager.sharedInstance.getLinphoneCore()
            let currentCall: OpaquePointer = linphone_core_get_current_call(lc)
            if linphone_core_get_calls_nb(lc) > 0 && linphone_call_get_state(currentCall) != LinphoneCallEnd {
                if(LinphoneManager.sharedInstance.isRinging) {
                    LinphoneManager.sharedInstance.stopRinging()
                }
                linphone_core_terminate_call(lc, currentCall)
                linphone_call_unref(currentCall)
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Terminate call..."
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Not calling"
                )
            }
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }
    
    @objc(sendMessage:) func sendMessage(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        self.commandDelegate.run(inBackground: {
            let lc : OpaquePointer = LinphoneManager.sharedInstance.getLinphoneCore()
            let pMessage = command.arguments[0] as? String ?? ""
            let pSipUri = command.arguments[1] as? String ?? ""
            let pNewChatConversation = command.arguments[2] as? Bool ?? false
            
            let address : OpaquePointer = linphone_core_interpret_url(lc, pSipUri)
            let chatRoom : OpaquePointer = linphone_core_get_chat_room(lc, address)
            
            if !pMessage.isEmpty && linphone_core_is_network_reachable(lc) == 1 {
                let message : OpaquePointer = linphone_chat_room_create_message(chatRoom, pMessage)
                linphone_chat_room_send_chat_message(chatRoom, message)
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: "Sending..."
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Error sending..."
                )
            }
            
            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        })
    }

    @objc(toggleSpeaker:) func toggleSpeaker(command: CDVInvokedUrlCommand) {

    }

    @objc(toggleMute:) func toggleMute(command: CDVInvokedUrlCommand) {

    }
    
    @objc(registerCallback:) func registerCallback(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            LinphoneManager.sharedInstance.setCallbackRegister(command: command, plugin: self)
        })
    }
    
    @objc(authInfoReqCallback:) func authInfoReqCallback(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            LinphoneManager.sharedInstance.setCallbackAuthInfo(command: command, plugin: self)
        })
    }
    
    @objc(callStateCallback:) func callStateCallback(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            LinphoneManager.sharedInstance.setCallbackCallState(command: command, plugin: self)
        })
    }
    
    @objc(callStatsCallback:) func callStatsCallback(command: CDVInvokedUrlCommand) {
        // TODO
    }
    
    @objc(globalCallback:) func globalCallback(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            LinphoneManager.sharedInstance.setCallbackGlobal(command: command, plugin: self)
        })
    }
    
    @objc(messageReceivedCallback:) func messageReceivedCallback(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            LinphoneManager.sharedInstance.setCallbackMessageReceived(command: command, plugin: self)
        })
    }

}
